---
title: "Intro to Epigenomics"
author: Wade Boohar
date: 11/03/24
updated: 03/07/24
---


```{r setup}
 knitr::opts_knit$set(root.dir = normalizePath("/home1/arsalans/analysis_data"))
```

Package Download and Data-cleaning
```{r}
if (!require("sesameData", quietly = TRUE))
BiocManager::install("sesameData")

if (!require("sesame", quietly = TRUE))
BiocManager::install("sesame")

if (!require("limma", quietly = TRUE))
BiocManager::install("limma")
install.packages("data.table")

```


Load in all necessary packages
```{r}
library(TCGAbiolinks)
library(sesame)
library(sesameData)
library(limma)
library(data.table)

```

```{r}
query <- GDCquery(project = "TCGA-BRCA",
                  data.category = "DNA Methylation",
                  data.type = "Methylation Beta Value",
                  platform = "Illumina Human Methylation 450",
                  data.format = "Dataframe")
GDCdownload(query)
methylation450 <- GDCprepare(query)
```

```{r}
methylation_clinical <- as.data.frame(methylation450@colData)
betas <- as.data.frame(methylation450@assays@data@listData)
cpg_sites <- as.data.frame(methylation450@rowRanges@elementMetadata)

column_mask <- ifelse(colnames(methylation_clinical) %in% c('treatments', 'primary_site', 'disease_type','sites_of_involvement'), F, T)

methylation_clinical <- methylation_clinical[,column_mask]
write.csv(methylation_clinical, 'brca_methylation_clinical.csv')


site_mask <- !grepl('-', cpg_sites$gene) & !grepl(';', cpg_sites$gene) & !is.na(cpg_sites$gene) & complete.cases(betas)
betas <- betas[site_mask,]
cpg_sites <- cpg_sites[site_mask,]

write.csv(betas, 'brca_methylation_betas.csv')
write.csv(cpg_sites, 'brca_cpg_sites.csv')
```

```{r}
methylation_clinical <- read.csv('brca_methylation_clinical.csv', row.names = 1)
betas <- read.csv('brca_methylation_betas.csv', row.names = 1)
cpg_sites <- read.csv('brca_cpg_sites.csv', row.names = 1)
```

```{r}
methylation_clinical <- read.csv("/project/rohs_1070/analysis_data/brca_methylation_clinical.csv", row.names = 1)

cpg_sites <- read.csv("/project/rohs_1070/analysis_data/brca_cpg_sites.csv", row.names = 1)


betas <- fread('/project/rohs_1070/analysis_data/brca_methylation_betas.csv', showProgress = TRUE)
betas <- as.data.frame(betas)
rownames(betas) <- betas[[1]] 
betas[[1]] <- NULL
# betas <- as.data.frame(betas)
# rownames(betas) <- betas[[1]] 
# betas[[1]] <- NULL 
```


(1) Naive Differential Methylation
```{r}

#masking out NAs
na_mask <- !is.na(methylation_clinical$age_at_diagnosis)
methylation_clinical <- methylation_clinical[na_mask,] 
betas_clean <- betas[, na_mask]

# Checking to see if sample #s match
dim(betas_clean)
nrow(methylation_clinical)

median <- median(methylation_clinical$age_at_diagnosis)

methylation_clinical$age_category <- ifelse(methylation_clinical$age_at_diagnosis >= median, "old", "young")

#fitting linear models using a "target value"
young_mask <- methylation_clinical$age_category == "young"

methylation_clinical$ages <- !young_mask

mval <- t(apply(betas_clean, 1, function(x) log2(x/(1-x)))) 
#mvalue is another statistic for methylation, centered at 0 and ranges from -1 to 1

design <- model.matrix(~ages, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)
```

```{r}
#Extracting model into dataframe
dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue =  -log10(p.adjust(fit2[["p.value"]][,2],method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)

#Visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
library(ggplot2)
ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")
ggsave("/home1/arsalans/490_cluster/week08_Epigenomics/EVPlot2.png")
```


(2) Direct comparison of methylation status to transcriptional activity

#...INSERT DESeq2 Stuff here to generate 'results'...
```{r}
library(DESeq2)

rna_query <- GDCquery(project = "TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)

rna_clinical <- as.data.frame(rna_se@colData)
rna_counts <- as.data.frame(rna_se@assays@data@listData$unstranded)
rna_genes <- as.data.frame(rna_se@rowRanges@elementMetadata)

#2
# Setting variables
rna_clinical <- as.data.frame(rna_se@colData) # clinical info
rna_counts <- as.data.frame(rna_se@assays@data$unstranded) # gene count
rna_genes <- as.data.frame(rna_se@rowRanges@elementMetadata) # 
rownames(rna_genes) <- rna_genes$gene_id # match gene_id
rownames(rna_counts) <- rna_genes$gene_id # match gene_id with rna_counts
colnames(rna_counts) <- rownames(rna_clinical) # match rows of rna_clinical with columns of rna_counts

#3
# renaming values for ease of reading
rna_clinical$shortLetterCode[rna_clinical$shortLetterCode=="TP"] <- "Tumor Primary"
rna_clinical$shortLetterCode[rna_clinical$shortLetterCode=="NT"] <- "Normal Tissue"
rna_clinical$shortLetterCode <- factor(rna_clinical$shortLetterCode)

#4
# removing NA values
sLCmask <- !is.na(rna_clinical$shortLetterCode) # this is for removing NA values in sLC
agemask <- !is.na(rna_clinical$age_at_index) # this is for removing NA values in age
rna_clinical <- rna_clinical[sLCmask & agemask, ]

#5
# align samples with each other
samemask  <- colnames(rna_counts) %in% rownames(rna_clinical)
rna_counts <- rna_counts[, samemask] ## this is currently outputting 0 variables which is not right (check #2) ##
rna_clinical <- rna_clinical[colnames(rna_counts), ]

#6
# this is the filter the genes and then realign them w rna_counts
keep_genes2 <- rowSums(rna_counts) >= 1000 # filter genes with >=1000
rna_counts <- rna_counts[keep_genes2, ]

dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                              colData = rna_clinical,
                              design = ~ definition)

dds_obj <- DESeq(dds)

results <- results(dds_obj, 
                   contrast = c("definition", "Primary solid Tumor", "Solid Tissue Normal"))

results
```

```{r}
# Add gene names properly - match by rownames
results$gene_name <- rna_genes[rownames(results), "gene_name"]

# Now filter for downregulated genes (log2FoldChange < 0, not < 3)
# Also make sure to remove NA values
downregulated <- results[!is.na(results$log2FoldChange) & 
                         results$log2FoldChange < 0, 'gene_name']

#you can also try looking at "upregulated" or "hypermethylated" !
downregulated <- results[(results$log2FoldChange < 3), 'gene_name']
hypomethylated <- dat[dat$foldchange < -1, 'geneName']
interest_genes <- intersect(downregulated, hypomethylated)
results@listData$gene_name
```

```{r}
# First, fix your code to use proper thresholds
# Add gene names properly - match by rownames
results$gene_name <- rna_genes[rownames(results), "gene_name"]

# Filter for downregulated genes (log2FoldChange < -1 and padj < 0.05)
downregulated <- results[!is.na(results$log2FoldChange) & 
                         !is.na(results$padj) &
                         results$log2FoldChange < -1 & 
                         results$padj < 0.05, 'gene_name']

# Filter for hypomethylated genes (foldchange < -1 and significant)
hypomethylated <- dat[!is.na(dat$foldchange) & 
                      !is.na(dat$logPvalue) &
                      dat$foldchange < -1 & 
                      dat$logPvalue >= -log10(0.05), 'geneName']

# Find intersection
interest_genes <- intersect(downregulated, hypomethylated)

# COUNT THE GENES
cat("Total downregulated genes (RNA-seq):", length(downregulated), "\n")
cat("Total hypomethylated genes (Methylation):", length(hypomethylated), "\n")
cat("Genes that are BOTH downregulated AND hypomethylated:", length(interest_genes), "\n\n")

# View the interest genes
cat("\n=== Downregulated & Hypomethylated Genes ===\n")
head(interest_genes, 30)

# Save to file
write.csv(interest_genes, "downregulated_hypomethylated_genes.csv", row.names = FALSE)
```



(Extra) Making Boxplots
```{r}
GENE<-"SCTR"

gene_counts_mask <- rna_genes$gene_name == GENE
gene_betas_mask <- cpg_sites$gene == GENE

rna_clinical_tumor <- rna_clinical$definition == "Primary solid Tumor"
methylation_clinical_tumor <- methylation_clinical$definition == "Primary solid Tumor"

rna_clinical_normal <- rna_clinical$definition == "Solid Tissue Normal"
methylation_clinical_normal <- methylation_clinical$definition == "Solid Tissue Normal"

rna_tumor <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_tumor])
methylation_tumor <- (betas[gene_betas_mask, methylation_clinical_tumor])

rna_normal <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_normal])
methylation_normal <- (betas[gene_betas_mask, methylation_clinical_normal])
```

```{r}
boxplot(rna_normal, rna_tumor, xlab='Group', ylab='Counts', names=c('Normal', 'Tumor'))
```
